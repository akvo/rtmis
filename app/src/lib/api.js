import axios from 'axios';

export const config = {
  baseURL: null,
  headers: {
    'Content-Type': 'application/json',
  },
};

const API = () => {
  const getConfig = () => {
    let current = config;
    if (api.baseURL) {
      current = {
        ...config,
        baseURL: api.baseURL,
      };
    }
    return api?.token
      ? {
          ...current,
          headers: { ...config.headers, Authorization: `Bearer ${api.token}` },
        }
      : current;
  };
  return {
    get: (url, config = {}) => axios({ url, ...getConfig(), ...config }),
    post: (url, data, config = {}) =>
      axios({
        url,
        method: 'POST',
        data,
        ...{
          ...getConfig(),
          headers: {
            ...getConfig().headers,
            ...config?.headers,
          },
        },
      }),
    put: (url, data, config) => axios({ url, method: 'PUT', data, ...getConfig(), ...config }),
    patch: (url, data, config) => axios({ url, method: 'PATCH', data, ...getConfig(), ...config }),
    delete: (url) => axios({ url, method: 'DELETE', ...getConfig() }),
    setToken: (token) => {
      api.token = token;
    },
    setServerURL: (serverURL) => {
      api.baseURL = serverURL;
    },
    getConfig,
  };
};

const api = API();

export default api;
