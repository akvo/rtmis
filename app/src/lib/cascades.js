import { Asset } from 'expo-asset';
import * as FileSystem from 'expo-file-system';
import * as SQLite from 'expo-sqlite';
import { conn, query } from '../database';

const DIR_NAME = 'SQLite';

const createSqliteDir = async () => {
  /**
   * Setup sqlite directory to save cascades sqlite from server
   */
  if (!(await FileSystem.getInfoAsync(FileSystem.documentDirectory + DIR_NAME)).exists) {
    await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + DIR_NAME);
  }
};

const download = (downloadUrl, fileUrl) => {
  const fileSql = fileUrl?.split('/')?.pop(); // get last segment as filename
  const pathSql = `${DIR_NAME}/${fileSql}`;
  FileSystem.getInfoAsync(FileSystem.documentDirectory + pathSql).then(({ exists }) => {
    if (!exists) {
      FileSystem.downloadAsync(downloadUrl, FileSystem.documentDirectory + pathSql);
    }
  });
};

const loadDataSource = async (source, id = null) => {
  const { file: cascadeName } = source;
  const db = SQLite.openDatabase(cascadeName);
  const readQuery = id ? query.read('nodes', { id }) : query.read('nodes');
  const readParam = id ? [id] : [];
  const result = await conn.tx(db, readQuery, readParam);
  return result;
};

const dropFiles = async () => {
  const Sqlfiles = await FileSystem.readDirectoryAsync(FileSystem.documentDirectory + DIR_NAME);
  Sqlfiles.forEach(async (file) => {
    if (file.includes('sqlite')) {
      const fileUri = FileSystem.documentDirectory + `${DIR_NAME}/${file}`;
      await FileSystem.deleteAsync(fileUri);
    }
  });
  return Sqlfiles;
};

export default cascades = {
  createSqliteDir,
  loadDataSource,
  download,
  dropFiles,
  DIR_NAME,
};
