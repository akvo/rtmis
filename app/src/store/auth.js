import { Store } from 'pullstate';
/**
 * Server config
 */
const AuthState = new Store({
  authenticationCode: '',
  useAuthenticationCode: false, // using code for authentication
  username: '',
  password: '',
  token: null,
});

export default AuthState;
