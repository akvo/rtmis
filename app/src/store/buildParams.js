import { Store } from 'pullstate';
import defaultBuildParams from '../build';

const BuildParamsState = new Store({
  authenticationType: defaultBuildParams?.authenticationType || [
    'code_assignment',
    'username',
    'password',
  ],
  serverURL: defaultBuildParams?.serverURL,
  debugMode: defaultBuildParams?.debugMode || false,
  dataSyncInterval: defaultBuildParams?.dataSyncInterval || 300,
  errorHandling: defaultBuildParams?.errorHandling || true,
  loggingLevel: defaultBuildParams?.loggingLevel || 'verbose',
  appVersion: defaultBuildParams?.appVersion || '1.0.0',
  gpsThreshold: defaultBuildParams?.gpsThreshold || 20,
});

export default BuildParamsState;
