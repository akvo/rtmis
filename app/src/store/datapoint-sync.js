import { Store } from 'pullstate';

const DatapointSyncState = new Store({
  inProgress: false,
  progress: 0,
  added: false,
  completed: false,
});

export default DatapointSyncState;
