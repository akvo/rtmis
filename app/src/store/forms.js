import { Store } from 'pullstate';

const FormState = new Store({
  form: {},
  currentValues: {}, // answers
  questionGroupListCurrentValues: {}, // answers for question group list component
  visitedQuestionGroup: [], // to store visited question group id
  surveyDuration: 0,
  surveyStart: null,
  cascades: {},
  lang: 'en',
});

export default FormState;
