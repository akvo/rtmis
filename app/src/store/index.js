export { default as AuthState } from './auth';
export { default as BuildParamsState } from './buildParams';
export { default as FormState } from './forms';
export { default as UIState } from './ui';
export { default as UserState } from './users';
