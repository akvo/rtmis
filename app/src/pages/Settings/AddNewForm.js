import React, { useState } from 'react';
import { BaseLayout } from '../../components';
import { View, StyleSheet, Platform, ToastAndroid } from 'react-native';
import { Input, Button, Text, Dialog } from '@rneui/themed';
import { i18n, api, cascades } from '../../lib';
import { UIState } from '../../store';
import { crudForms } from '../../database/crud';

const AddNewForm = ({ navigation }) => {
  const { online: isNetworkAvailable, lang: activeLang } = UIState.useState((s) => s);
  const trans = i18n.text(activeLang);
  const [loading, setLoading] = useState(false);
  const [formId, setFormId] = useState(null);
  const [error, setError] = useState(null);

  const handleDownloadForm = () => {
    if (!isNetworkAvailable) {
      if (Platform.OS === 'android') {
        ToastAndroid.show(trans.authErrorNoConn, ToastAndroid.LONG);
      }
      return;
    }
    setLoading(true);
    setError(null);
    api
      .get(`/forms/${formId}`)
      .then(async (res) => {
        try {
          const { data } = res;
          await cascades.createSqliteDir();
          // save forms
          const savedForm = await crudForms.addForm({
            id: data.id,
            version: data.version,
            formJSON: data,
          });
          console.info('Saved Forms...', data.id, savedForm);
          // download cascades files
          if (data?.cascades?.length) {
            data.cascades.forEach((cascadeFile) => {
              const downloadUrl = api.getConfig().baseURL + cascadeFile;
              cascades.download(downloadUrl, cascadeFile);
            });
          }
          setTimeout(() => {
            navigation.navigate('Home');
          }, 100);
        } catch (err) {
          console.error(err);
        }
      })
      .catch((err) => {
        const { status: errStatus } = err?.response;
        if ([400, 401].includes(errStatus)) {
          setError(trans.authErrorPasscode);
        } else {
          setError(err?.message);
        }
      })
      .finally(() => setLoading(false));
  };

  return (
    <BaseLayout title={trans.settingAddNewFormPageTitle} rightComponent={false}>
      <BaseLayout.Content>
        <View style={styles.container}>
          <Input
            placeholder={trans.inputFormID}
            testID="input-form-id"
            autoFocus
            value={formId}
            onChangeText={setFormId}
            keyboardType="numeric"
            inputContainerStyle={styles.inputFormId}
          />
          <Button testID="button-download-form" onPress={handleDownloadForm}>
            {trans.downloadFormButton}
          </Button>
          {error && (
            <Text style={styles.errorText} testID="fetch-error-text">
              {error}
            </Text>
          )}
        </View>
      </BaseLayout.Content>
      {/* Loading dialog */}
      <Dialog isVisible={loading} style={styles.dialogLoadingContainer}>
        <Dialog.Loading />
        <Text style={styles.dialogLoadingText}>{trans.fetchingData}</Text>
      </Dialog>
    </BaseLayout>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    paddingHorizontal: 16,
    paddingTop: 24,
  },
  inputFormId: {
    width: '100%',
  },
  errorText: { color: 'red', fontStyle: 'italic', marginHorizontal: 10, marginTop: 8 },
  dialogLoadingContainer: {
    flex: 1,
  },
  dialogLoadingText: {
    textAlign: 'center',
    fontStyle: 'italic',
  },
});

export default AddNewForm;
