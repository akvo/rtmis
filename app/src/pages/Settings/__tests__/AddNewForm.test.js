import React from 'react';
import renderer from 'react-test-renderer';
import { render, renderHook, fireEvent, act, waitFor } from '@testing-library/react-native';
import { useNavigation } from '@react-navigation/native';
import { Platform } from 'react-native';
import api from '../../../lib/api';
import cascades from '../../../lib/cascades';

import AddNewForm from '../AddNewForm';
import { UIState } from '../../../store';

jest.mock('../../../lib/api');
jest.mock('../../../lib/cascades');
jest.mock('../../../database/crud');

describe('AddNewForm Page', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<AddNewForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show input form id and download button', () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;

    const wrapper = render(<AddNewForm navigation={navigation} />);

    const fidInput = wrapper.getByTestId('input-form-id');
    expect(fidInput).toBeDefined();
    const downloadButton = wrapper.getByTestId('button-download-form');
    expect(downloadButton).toBeDefined();
  });

  it('should not fetch form if network not available', async () => {
    Platform.OS = 'android';
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.post.mockImplementation(() => Promise.resolve({ data: { formsUrl: [] } }));

    const wrapper = render(<AddNewForm navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = false;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).not.toHaveBeenCalled();
    });
  });

  it('should fetch form if network available', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.get.mockImplementation(() => Promise.resolve({ data: { formsUrl: [] } }));

    const wrapper = render(<AddNewForm navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).toHaveBeenCalledWith('/forms/1');
    });
  });

  it('it should navigate to home page after form downloaded', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    // url: /forms/1
    const FormId1 = {
      id: 1,
      name: 'Household',
      version: '1.0.0',
      cascades: ['/cascades/1.sqlite'],
      question_group: [],
    };
    // url: /sqlite/file.sqlite
    const mockFile = 'file.sqlite';

    api.getConfig.mockImplementation(() => ({ baseURL: 'http://example.com' }));
    api.get.mockImplementation((url) => {
      if (url === '/forms/1') {
        return Promise.resolve({ data: FormId1 });
      }
      if (url === '/cascades/1.sqlite') {
        return Promise.resolve({ data: mockFile });
      }
    });

    const wrapper = render(<AddNewForm navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).toHaveBeenCalledWith('/forms/1');
      expect(cascades.download).toHaveBeenCalledWith(
        'http://example.com/cascades/1.sqlite',
        '/cascades/1.sqlite',
      );
    });
    await waitFor(() => expect(navigation.navigate).toHaveBeenCalledWith('Home'));
  });

  it('should show error text if fetch form error', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const mockErrorData = { message: 'Failed' };
    api.get.mockImplementation(() =>
      Promise.reject({ response: { ...mockErrorData, status: 400 } }),
    );

    const wrapper = render(<AddNewForm navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).toHaveBeenCalledWith('/forms/1');
      const errorText = wrapper.getByTestId('fetch-error-text');
      expect(errorText).toBeTruthy();
    });
  });

  it('should be error 500', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.get.mockImplementation(() =>
      Promise.reject({ response: { message: 'Failed', status: 500 } }),
    );

    const wrapper = render(<AddNewForm navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).toHaveBeenCalledWith('/forms/1');
    });
  });
});
