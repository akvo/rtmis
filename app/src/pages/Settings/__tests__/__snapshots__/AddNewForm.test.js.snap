// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewForm Page renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9fafb",
      "flex": 1,
      "flexDirection": "column",
      "flexWrap": "wrap",
    }
  }
  testID="stack-container"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "minheight": 78,
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="base-layout-page-title"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontSize": 18,
              "paddingTop": 4,
              "textAlign": "center",
            }
          }
          testID="page-title"
        >
          Add New Form
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      />
    </RNCSafeAreaView>
  </View>
  <View
    style={
      {
        "backgroundColor": "#f9fafb",
        "flex": 1,
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
    testID="stack-container"
  >
    <View
      style={
        {
          "paddingHorizontal": 16,
          "paddingTop": 24,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
        testID="RNE__Input__view-wrapper"
      >
        <View
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "transform": [
                {
                  "translateX": 0,
                },
              ],
              "width": "100%",
            }
          }
        >
          <TextInput
            autoFocus={true}
            editable={true}
            keyboardType="numeric"
            onChangeText={[Function]}
            placeholder="Form ID"
            placeholderTextColor="#86939e"
            style={
              {
                "color": "#242424",
                "flex": 1,
                "fontSize": 18,
                "minHeight": 40,
              }
            }
            testID="input-form-id"
            underlineColorAndroid="transparent"
            value={null}
          />
        </View>
        <Text
          style={
            {
              "color": "#ff190c",
              "fontSize": 12,
              "margin": 5,
            }
          }
        />
      </View>
      <View
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 2,
            },
            undefined,
            false,
          ]
        }
        testID="RNE_BUTTON_WRAPPER"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": false,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="button-download-form"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#2089dc",
                "borderColor": "#2089dc",
                "borderRadius": 2,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
                "paddingHorizontal": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Download Form
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    style={
      {
        "flex": 1,
        "width": "100%",
      }
    }
    testID="Internal__Overlay"
    transparent={true}
    visible={false}
  />
</View>
`;
